### PR TITLE
1193 read price data

### PR DESCRIPTION
**Pull request by jacobvjk**
closes ADO 1193: https://dev.azure.com/2DegreesInvesting/2DegreesInvesting/_workitems/edit/1193

This PR:
- adds a function read_price_data() that safely reads in market prices for further calculation in the transition risk stress test
- adds two child functions, one covering the old structure of prices read_price_data_internal_old(), one covering the new long format read_price_data_internal().
- idea is to deprecate the old version in the medium run
- call new read_price_data function in main scripts

### CODE REVIEW

File: R/read_price_data_internal.R [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  expected_columns <- c(
+    "year", "source", "scenario", "scenario_geography", "technology",
+    "indicator", "unit", "price"
+  )
+
+  data_has_expected_columns <- all(expected_columns %in% colnames(data))
+  stopifnot(data_has_expected_columns)
+
+  return(data)
+  # TODO: add check that looks if all technologies are given (requires new input arg)
```
Comment (MirijaHa) [inline R/read_price_data_internal.R:35]: follow up?

File: R/read_price_data_internal_old.R [hunk 2]
```diff
@@ -0,0 +1,41 @@
+#' Read in price data from csv and check that all expected columns are given.
+#'
+#' @description This function reads in price data using the old wide data format.
```
Comment (MirijaHa) [inline R/read_price_data_internal_old.R:3]: is it possible to use a more concrete description than old wide data :) ? E.g. is it a particular year or source that will work with this?

File: stress_test_model_eq_cb.R [hunk 3]
```diff
@@ -409,7 +412,10 @@ scenario_data <- scenario_data %>%
       scenario_geography == scenario_geography_filter)
 
 # Load price data----------------------------------------
-df_price <- readr::read_csv(file.path(data_location, paste0("prices_data_", price_data_version, ".csv")), col_types = "ncccccncncncnc") %>%
+df_price <- read_price_data(
+    path = file.path(data_location, paste0("prices_data_", price_data_version, ".csv")),
+    version = "old"
```
Comment (MirijaHa) [inline stress_test_model_eq_cb.R:417]: I saw that version is always old though price_data_version differs between the 3 calls of the function. Is that intentional?

File: web_tool_stress_test.R [hunk 4]
```diff
@@ -221,7 +224,10 @@ scenario_data <- scenario_data %>%
 
 # %>% filter(year %in% c(start_year,2020, 2021, 2022, 2023, 2024, 2025, 2030, 2035, 2040))
 
-df_price <- readr::read_csv(file.path(data_location, glue::glue("prices_data_{price_data_version}.csv")), col_types = cols()) %>%
+df_price <- read_price_data(
+    path = file.path(data_location, paste0("prices_data_", price_data_version, ".csv")),
+    version = "old"
+  ) %>%
   filter(year >= start_year) %>%
```
Comment (MirijaHa) [inline web_tool_stress_test.R:231]: out of interest, where does start_year come from here? I did not find an assignment and assume it is somehow made available in he env when calling the script?

File: tests/testthat/test-read_price_data.R [hunk 5]
```diff
@@ -0,0 +1,100 @@
+test_that("without specified arguments, read_price_data() throws error", {
```
Comment (MirijaHa) [inline tests/testthat/test-read_price_data.R:1]: This test and the next actually tests not our code but the functioning of R base,thus the tests are superfluous, I would rather remove them

**Review by MirijaHa — COMMENTED** [review]
I would suggest to put all 3 read functions in one file. Mostly because Hadley says so (i.e. both extremes a file per function, and 1 file with all functions are not recommendable).
Also I wonder if the section in readme on prices data is still up to date now?

File: tests/testthat/test-read_price_data.R (hunk 5 above)
Comment (jacobvjk) [reply tests/testthat/test-read_price_data.R:1]: Fair enough, I guess it's hard to argue that my intention was to test R base.

Comment (MirijaHa) [reply tests/testthat/test-read_price_data.R:1]: that would qualify as hybris

File: stress_test_model_eq_cb.R (hunk 3 above)
Comment (jacobvjk) [reply stress_test_model_eq_cb.R:409]: yes, only price_data_version == 2021Q3 would be version new as of now. It is currently not being used anywhere

File: R/read_price_data_internal_old.R (hunk 2 above)
Comment (jacobvjk) [reply R/read_price_data_internal_old.R]: I wish I had a way to verify the exact year the data is taken from. But it just existed when I joined...

File: R/read_price_data_internal.R (hunk 1 above)
Comment (jacobvjk) [reply R/read_price_data_internal.R]: Hm yes, I'm still a bit undecided if that should be in this function..

File: web_tool_stress_test.R (hunk 4 above)
Comment (jacobvjk) [reply web_tool_stress_test.R:229]: project configuration

File: R/read_price_data.R [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+#' @param version A string indicating whether to prepare the capacity factors
+#'   based on the old price data structure or the new long format. Must be
+#'   either "old" or "new".
+#'
+#' @family import functions
+#'
+#' @export
+read_price_data <- function(path,
+                            version) {
+  valid_input_file_path <- file.exists(file.path(path))
```
[the remark this answers is not in the document]
Comment (jacobvjk) [reply R/read_price_data.R]: added

File: R/read_price_data_internal.R (hunk 1 above)
Comment (jacobvjk) [reply R/read_price_data_internal.R]: integrated

File: R/read_price_data.R [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+      NPS = "d",
+      nps_source = "c",
+      SDS = "d",
+      sds_source = "c",
+      Baseline = "d",
+      baseline_source = "c"
+    )
+  )
+
+  data %>% validate_data_has_expected_cols(
```
Comment (MirijaHa) [inline R/read_price_data.R:102]: NP: style

File: R/read_price_data.R [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+      scenario = "c",
+      scenario_geography = "c",
+      technology = "c",
+      indicator = "c",
+      unit = "c",
+      price = "d"
+    )
+  )
+
+  data %>% validate_data_has_expected_cols(
```
Comment (MirijaHa) [inline R/read_price_data.R:61]: NP (nitpicking): style

File: tests/testthat/test-read_price_data.R [hunk 9]
```diff
@@ -0,0 +1,191 @@
+# test read_price_data()
+test_that("with missing argument for version, read_price_data() throws error", {
```
Comment (MirijaHa) [inline tests/testthat/test-read_price_data.R:2]: this is still a test testing base R

**Review by MirijaHa — COMMENTED** [review]
2 small things, otherwise LGTM

File: tests/testthat/test-read_price_data.R (hunk 9 above)
Comment (jacobvjk) [reply tests/testthat/test-read_price_data.R:2]: It's Monday

**Review by MirijaHa — APPROVED** [review]
Beautiful

**Pull request merged**